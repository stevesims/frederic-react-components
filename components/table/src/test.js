import React from 'react';
import { mount, shallow } from 'enzyme';
import { createMatchers } from 'jest-emotion';
import * as emotion from 'emotion';

import Table from '.';

expect.extend(createMatchers(emotion));

const titles = ['Heading 1', 'Heading 2', 'Heading 3', 'Heading 4'];
const rows = [
  ['Content 1-1', 'Content 1-2', 'Content 1-3', 'Content 1-4'],
  ['Content 2-1', 'Content 2-2', 'Content 2-3', 'Content 2-4'],
  ['Content 3-1', 'Content 3-2', 'Content 3-3', 'Content 3-4'],
];
const columnTableNames = ['one', 'two', 'three', ['i', 'am', 'named', 'individually']];
const rowTableNamesWithTitles = ['heading', 'one', ['i', 'am', 'named', 'individually'], 'three'];
const rowTableNames = ['one', ['i', 'am', 'named', 'individually'], 'three'];

describe('Table', () => {
  it('renders with only required props', () => {
    const wrapper = shallow(<Table rows={rows} />);
    expect(wrapper.exists()).toBe(true);
  });

  it('renders rows from prop', () => {
    const wrapper = mount(<Table rows={rows} />);
    expect(wrapper.find('tr')).toHaveLength(3);
  });

  it('renders titles from prop', () => {
    const wrapper = mount(<Table rows={rows} titles={titles} />);
    expect(wrapper.find('TableHeading')).toHaveLength(4);
  });

  it('names each cell according to its column', () => {
    const wrapper = mount(
      <Table rows={rows} titles={titles} name="name" names={columnTableNames} />);
    const th = wrapper.find('TableHeading').at(2);
    expect(th.prop('name')).toBe('three');
    const td = wrapper.find('TableData').at(2);
    expect(td.prop('name')).toBe('three');
  });

  it('renders rows with header column', () => {
    const wrapper = mount(<Table
      rows={rows}
      name="name"
      titles={titles}
      names={columnTableNames}
      rowIncludesHeading
      nameByRow
    />);
    expect(wrapper.find('TableHeading')).toHaveLength(7);
  });

  it('names each cell according to its row, no titles', () => {
    const wrapper = mount(<Table
      rows={rows}
      name="name"
      names={rowTableNames}
      rowIncludesHeading
      nameByRow
    />);

    let th = wrapper.find('TableHeading').at(0);
    expect(th.prop('name')).toBe('one');

    let td = wrapper.find('TableData').at(0);
    expect(td.prop('name')).toBe('one');

    th = wrapper.find('TableHeading').at(1);
    expect(th.prop('name')).toBe('i');

    td = wrapper.find('TableData').at(3);
    expect(td.prop('name')).toBe('am');

    td = wrapper.find('TableData').at(4);
    expect(td.prop('name')).toBe('named');

    td = wrapper.find('TableData').at(5);
    expect(td.prop('name')).toBe('individually');

    th = wrapper.find('TableHeading').at(2);
    expect(th.prop('name')).toBe('three');

    td = wrapper.find('TableData').at(8);
    expect(td.prop('name')).toBe('three');
  });

  it('renders name on table element', () => {
    const wrapper = mount(<Table
      name="tableNameAtt"
      rows={rows}
      rowIncludesHeading
      nameByRow
    />);

    expect(wrapper.find('table').prop('name')).toBe('tableNameAtt');
  });

  it('names each cell according to its row, with titles', () => {
    const wrapper = mount(<Table
      rows={rows}
      name="name"
      names={rowTableNamesWithTitles}
      rowIncludesHeading
      nameByRow
      titles={titles}
    />);

    let th = wrapper.find('TableHeading').at(3);
    expect(th.prop('name')).toBe('heading');

    th = wrapper.find('TableHeading').at(4);
    expect(th.prop('name')).toBe('one');

    let td = wrapper.find('TableData').at(0);
    expect(td.prop('name')).toBe('one');

    th = wrapper.find('TableHeading').at(5);
    expect(th.prop('name')).toBe('i');

    td = wrapper.find('TableData').at(3);
    expect(td.prop('name')).toBe('am');

    td = wrapper.find('TableData').at(4);
    expect(td.prop('name')).toBe('named');

    td = wrapper.find('TableData').at(5);
    expect(td.prop('name')).toBe('individually');

    th = wrapper.find('TableHeading').at(6);
    expect(th.prop('name')).toBe('three');

    td = wrapper.find('TableData').at(8);
    expect(td.prop('name')).toBe('three');
  });

  it('sets th width according to rowHeading prop', () => {
    const wrapper = mount(<Table
      rows={rows}
      name="name"
      names={rowTableNamesWithTitles}
      rowIncludesHeading
      nameByRow
      titles={titles}
    />);
    expect(wrapper.find('TableHeading').at(3)).not.toHaveStyleRule('width');
    const t = titles.slice(-1);
    const r = rows.map(row => row.slice(-1));
    wrapper.setProps({ titles: t, rows: r });
    expect(wrapper.find('TableHeading').at(3)).toHaveStyleRule('width', '25%');
  });

  it('sets table-layout according to flexibleColumns prop', () => {
    const wrapper = mount(<Table
      rows={rows}
      name="name"
      names={rowTableNamesWithTitles}
      rowIncludesHeading
      nameByRow
      titles={titles}
    />);
    expect(wrapper).toHaveStyleRule('table-layout', 'fixed');
    wrapper.setProps({ flexibleColumns: true });
    expect(wrapper).toHaveStyleRule('table-layout', 'auto');
  });

  it('matches snapshot', () => {
    const wrapper = mount(<Table
      rows={rows}
      name="name"
      names={rowTableNamesWithTitles}
      rowIncludesHeading
      flexibleColumns
      nameByRow
      titles={titles}
    />);
    expect(wrapper).toMatchSnapshot();
  });
});
