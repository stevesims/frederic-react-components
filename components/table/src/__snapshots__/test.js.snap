// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table matches snapshot 1`] = `
.emotion-34 {
  margin-bottom: 15px;
}

@media only screen and (min-width:641px) {
  .emotion-34 {
    margin-bottom: 15px;
  }
}

.emotion-10 {
  border-bottom: 1px solid #bfc1c3;
  display: table-cell;
  font-size: 14px;
  padding: 15px 4px;
  vertical-align: top;
}

.emotion-10:first-child {
  padding: 15px 4px 15px 0;
}

.emotion-10:last-child {
  padding: 15px 0 15px 4px;
}

.emotion-0 {
  border-bottom: 1px solid #bfc1c3;
  display: table-cell;
  font-size: 14px;
  padding: 15px 4px;
  vertical-align: top;
  font-weight: bold;
  text-align: left;
}

.emotion-0:first-child {
  padding: 15px 4px 15px 0;
}

.emotion-0:last-child {
  padding: 15px 0 15px 4px;
}

.emotion-32 {
  display: table;
  font-family: "nta",Arial,sans-serif;
  width: 100%;
  table-layout: auto;
  margin-bottom: 15px;
}

@media only screen and (min-width:641px) {
  .emotion-32 {
    margin-bottom: 15px;
  }
}

<Styled(Table)
  flexibleColumns={true}
  name="name"
  nameByRow={true}
  names={
    Array [
      "heading",
      "one",
      Array [
        "i",
        "am",
        "named",
        "individually",
      ],
      "three",
    ]
  }
  rowIncludesHeading={true}
  rows={
    Array [
      Array [
        "Content 1-1",
        "Content 1-2",
        "Content 1-3",
        "Content 1-4",
      ],
      Array [
        "Content 2-1",
        "Content 2-2",
        "Content 2-3",
        "Content 2-4",
      ],
      Array [
        "Content 3-1",
        "Content 3-2",
        "Content 3-3",
        "Content 3-4",
      ],
    ]
  }
  titles={
    Array [
      "Heading 1",
      "Heading 2",
      "Heading 3",
      "Heading 4",
    ]
  }
>
  <Table
    className="emotion-34"
    flexibleColumns={true}
    name="name"
    nameByRow={true}
    names={
      Array [
        "heading",
        "one",
        Array [
          "i",
          "am",
          "named",
          "individually",
        ],
        "three",
      ]
    }
    rowIncludesHeading={true}
    rows={
      Array [
        Array [
          "Content 1-1",
          "Content 1-2",
          "Content 1-3",
          "Content 1-4",
        ],
        Array [
          "Content 2-1",
          "Content 2-2",
          "Content 2-3",
          "Content 2-4",
        ],
        Array [
          "Content 3-1",
          "Content 3-2",
          "Content 3-3",
          "Content 3-4",
        ],
      ]
    }
    titles={
      Array [
        "Heading 1",
        "Heading 2",
        "Heading 3",
        "Heading 4",
      ]
    }
  >
    <TableContainer
      className="emotion-34"
      flexibleColumns={true}
    >
      <table
        className="emotion-32 emotion-33"
      >
        <thead>
          <tr>
            <TableHeading
              key="0"
              name="heading"
            >
              <th
                className="emotion-0 emotion-1"
                name="heading"
              >
                Heading 1
              </th>
            </TableHeading>
            <TableHeading
              key="1"
              name="heading"
            >
              <th
                className="emotion-0 emotion-1"
                name="heading"
              >
                Heading 2
              </th>
            </TableHeading>
            <TableHeading
              key="2"
              name="heading"
            >
              <th
                className="emotion-0 emotion-1"
                name="heading"
              >
                Heading 3
              </th>
            </TableHeading>
            <TableHeading
              key="3"
              name="heading"
            >
              <th
                className="emotion-0 emotion-1"
                name="heading"
              >
                Heading 4
              </th>
            </TableHeading>
          </tr>
        </thead>
        <tbody>
          <tr
            key="0"
          >
            <TableHeading
              columnCount={4}
              key="0"
              name="one"
              rowHeading={true}
            >
              <th
                className="emotion-0 emotion-1"
                name="one"
              >
                Content 1-1
              </th>
            </TableHeading>
            <TableData
              key="1"
              name="one"
            >
              <td
                className="emotion-10 emotion-11"
                name="one"
              >
                Content 1-2
              </td>
            </TableData>
            <TableData
              key="2"
              name="one"
            >
              <td
                className="emotion-10 emotion-11"
                name="one"
              >
                Content 1-3
              </td>
            </TableData>
            <TableData
              key="3"
              name="one"
            >
              <td
                className="emotion-10 emotion-11"
                name="one"
              >
                Content 1-4
              </td>
            </TableData>
          </tr>
          <tr
            key="1"
          >
            <TableHeading
              columnCount={4}
              key="0"
              name="i"
              rowHeading={true}
            >
              <th
                className="emotion-0 emotion-1"
                name="i"
              >
                Content 2-1
              </th>
            </TableHeading>
            <TableData
              key="1"
              name="am"
            >
              <td
                className="emotion-10 emotion-11"
                name="am"
              >
                Content 2-2
              </td>
            </TableData>
            <TableData
              key="2"
              name="named"
            >
              <td
                className="emotion-10 emotion-11"
                name="named"
              >
                Content 2-3
              </td>
            </TableData>
            <TableData
              key="3"
              name="individually"
            >
              <td
                className="emotion-10 emotion-11"
                name="individually"
              >
                Content 2-4
              </td>
            </TableData>
          </tr>
          <tr
            key="2"
          >
            <TableHeading
              columnCount={4}
              key="0"
              name="three"
              rowHeading={true}
            >
              <th
                className="emotion-0 emotion-1"
                name="three"
              >
                Content 3-1
              </th>
            </TableHeading>
            <TableData
              key="1"
              name="three"
            >
              <td
                className="emotion-10 emotion-11"
                name="three"
              >
                Content 3-2
              </td>
            </TableData>
            <TableData
              key="2"
              name="three"
            >
              <td
                className="emotion-10 emotion-11"
                name="three"
              >
                Content 3-3
              </td>
            </TableData>
            <TableData
              key="3"
              name="three"
            >
              <td
                className="emotion-10 emotion-11"
                name="three"
              >
                Content 3-4
              </td>
            </TableData>
          </tr>
        </tbody>
      </table>
    </TableContainer>
  </Table>
</Styled(Table)>
`;
